### ISSUE TITLE
DeepSkyStacker 4.2.4 Beta 4 crashes trying to select a comet

### ISSUE THREAD

**Issue by FelixJongleur42** [deepskystacker/DSS#105]
DeepSkyStacker 4.2.4 Beta 4 crashes trying to select a comet ("comet selection mode"). There is no crash screen however.
Let me know how I can debug this. I will try to build the software from source I guess.

**Comment by perdrix52** [comment]
Merde!  Yes I suspect you may have to build and debug that.

**Comment by FelixJongleur42** [comment]
Ok I will try to setup Visual Studio and such.

**Comment by FelixJongleur42** [comment]
Well it looks like a off-by-one-error, with some assertion failing (-1 vs. 0).

```
 	DeepSkyStacker.exe!ZException::assertParameter(const char * exceptionText, ZExceptionLocation location) Zeile 614	C++
 	DeepSkyStacker.exe!CGrayBitmapT<double>::CheckXY(long x, long y) Zeile 1725	C++
 	DeepSkyStacker.exe!CGrayBitmapT<double>::GetValue(long i, long j, double & fGray) Zeile 2094	C++
>	DeepSkyStacker.exe!CGrayBitmapT<unsigned short>::GetPixel(long i, long j, double & fGray) Zeile 2234	C++
 	DeepSkyStacker.exe!CRegisteredFrame::RegisterSubRect(CMemoryBitmap * pBitmap, CRect & rc) Zeile 504	C++
 	DeepSkyStacker.exe!CEditStarsSink::DetectStars(const CPointExt & pt, CRect & rcCheck, std::vector<CStar,std::allocator<CStar>> & vStars) Zeile 673	C++
 	DeepSkyStacker.exe!CEditStarsSink::GetOverlayImage(CRect & rcClient) Zeile 1076	C++
 	DeepSkyStacker.exe!CWndImage::CreateBufferBitmap() Zeile 514	C++
 	DeepSkyStacker.exe!CWndImage::RefreshOverlay() Zeile 1191	C++
 	DeepSkyStacker.exe!CWndImage::OnMouseMove(unsigned int nFlags, CPoint point) Zeile 1060	C++
```
RegisterEngine.cpp

```
								for (k = 0;k<8;k++)
								{
									LONG		ldX = vPixels[k].m_lXDir*r;
									LONG		ldY = vPixels[k].m_lYDir*r;

									pBitmap->GetPixel(i+ldX, j+ldY, vPixels[k].m_fIntensity);
								};
```

ldY is -22 and j is 21, so the second argument to GetPixel is -1, which leads to the assertion.
That is probably all I can do for now in order to narrow it down. The image happens to have 22 stars if this helps (probably just a coincidence).

**Comment by perdrix52** [comment]
Thank you for the initial debug work most helpful.   Unfortunately I couldn't spot the bug immediately.  So this has to go onto the "to do" list.

PS how were you trying to select the comet nucleus - should be a simple single click, not a click and drag ...

David

**Comment by FelixJongleur42** [comment]
Thanks for the quick response, David, I was just hovering with the mouse (no clicks yet).

**Comment by perdrix52** [comment]
Very odd - I tried to reproduce the problem here and totally failed.  Please could I ask you to email me the light file that I can use to reproduce this?

**Comment by FelixJongleur42** [comment]
I did some debug output, somehow the radius for some subrect is getting to big and GetPixel asserts.
It asserts as soon as I hover around the comet.

![IMG_1033](https://user-images.githubusercontent.com/19993980/87470597-4871f300-c61d-11ea-9cb2-103b25e6f79f.JPG)
[IMG_1033.Info.txt](https://github.com/deepskystacker/DSS/files/4921024/IMG_1033.Info.txt)

```
D:\florian\OneDrive\Documents\GitHub\DSS\DeepSkyStacker\RegisterEngine.cpp(505) : r=19, k=5, ldX=+19, ldX=+19, GetPixel(+71, +39, ...)
D:\florian\OneDrive\Documents\GitHub\DSS\DeepSkyStacker\RegisterEngine.cpp(505) : r=19, k=6, ldX=-19, ldX=+19, GetPixel(+33, +39, ...)
D:\florian\OneDrive\Documents\GitHub\DSS\DeepSkyStacker\RegisterEngine.cpp(505) : r=19, k=7, ldX=-19, ldX=-19, GetPixel(+33, +01, ...)
D:\florian\OneDrive\Documents\GitHub\DSS\DeepSkyStacker\RegisterEngine.cpp(505) : r=20, k=0, ldX=+00, ldX=-20, GetPixel(+52, +00, ...)
D:\florian\OneDrive\Documents\GitHub\DSS\DeepSkyStacker\RegisterEngine.cpp(505) : r=20, k=1, ldX=+20, ldX=+00, GetPixel(+72, +20, ...)
D:\florian\OneDrive\Documents\GitHub\DSS\DeepSkyStacker\RegisterEngine.cpp(505) : r=20, k=2, ldX=+00, ldX=+20, GetPixel(+52, +40, ...)
D:\florian\OneDrive\Documents\GitHub\DSS\DeepSkyStacker\RegisterEngine.cpp(505) : r=20, k=3, ldX=-20, ldX=+00, GetPixel(+32, +20, ...)
D:\florian\OneDrive\Documents\GitHub\DSS\DeepSkyStacker\RegisterEngine.cpp(505) : r=20, k=4, ldX=+20, ldX=-20, GetPixel(+72, +00, ...)
D:\florian\OneDrive\Documents\GitHub\DSS\DeepSkyStacker\RegisterEngine.cpp(505) : r=20, k=5, ldX=+20, ldX=+20, GetPixel(+72, +40, ...)
D:\florian\OneDrive\Documents\GitHub\DSS\DeepSkyStacker\RegisterEngine.cpp(505) : r=20, k=6, ldX=-20, ldX=+20, GetPixel(+32, +40, ...)
D:\florian\OneDrive\Documents\GitHub\DSS\DeepSkyStacker\RegisterEngine.cpp(505) : r=20, k=7, ldX=-20, ldX=-20, GetPixel(+32, +00, ...)
D:\florian\OneDrive\Documents\GitHub\DSS\DeepSkyStacker\RegisterEngine.cpp(505) : r=21, k=0, ldX=+00, ldX=-21, GetPixel(+52, -01, ...)
```

Modified code:
```
#include <crtdbg.h>
...
// around line 505
								for (k = 0;k<8;k++)
								{
									LONG		ldX = vPixels[k].m_lXDir*r;
									LONG		ldY = vPixels[k].m_lYDir*r;
									_RPTFN(_CRT_WARN, "r=%02d, k=%d, ldX=%+03d, ldX=%+03d, GetPixel(%+03d, %+03d, ...)\n", r, k, ldX, ldY, (i + ldX), (j + ldY));
									pBitmap->GetPixel(i+ldX, j+ldY, vPixels[k].m_fIntensity);
								};
```

**Comment by FelixJongleur42** [comment]
Just tried to go back to a version where it would work, and I noticed that I cannot select the comet in either version.
Am I misusing DSS?

All other versions do not crash, though 4.2.3, 4.2.2, 4.2.1, 4.1.1, 4.1.0

**Comment by perdrix52** [comment]
Not crashing in earlier versions - quite possible as I added a lot of checking code recently to try to catch other problems.   Maybe I got this test incorrect?  Until I can reproduce it however ...

As for using DSS incorrectly - even it you were, it shouldn't crash on you!

**Comment by perdrix52** [comment]
If yo can't select the comet by just clicking on the heart of the nucleus (because DSS hasn't detected it), I think you can "Shift-Click" to force the selection.

**Comment by perdrix52** [comment]
Please could you upload at least one of your lights to somewhere like DropBox or GoogleDrive or similar and let me know where I can DL them from.

Thanks
David

**Comment by FelixJongleur42** [comment]
Is the attach under https://github.com/deepskystacker/DSS/issues/105#issuecomment-658384769
Ok or do you need more?

**Comment by perdrix52** [comment]
I agree there's a problem I too had DSS crash on me with that image.   So will be able to investigate the bug. 

However, AFAICT there are no stars in the image, just the comet.   If you try to register the frame it just finds a few "hot pixels".    This means you won't be able to stack your images using DSS.  The process in DSS is to register all your light frames to locate the stars, and then to mark the comet in the first and last frames if you want to do comet stacking.

**Comment by perdrix52** [comment]
I've fixed the bug (I think) and the fix will ship in beta 5

David

**Comment by perdrix52** [comment]
Fixed 4.2.4

**Comment by FelixJongleur42** [comment]
Great! Thank you!
